File: matrix_class/src/NewtonDividedDifference.cpp
#include "../include/NewtonDividedDifference.hpp"
#include <cmath>
#include <iomanip>
#include <iostream>
using namespace std;

// default constructor
NewtonDividedDifference::NewtonDividedDifference()
    : Interpolation(), computed(false) {}

// parameterized constructor — takes Matrix objects
NewtonDividedDifference::NewtonDividedDifference(const Matrix &x, const Matrix &y)
    : Interpolation(x, y), computed(false) {}

// build the divided difference table
// stores as an n×n lower-triangular Matrix where:
//   ddTable(i, 0) = f[xᵢ] = yᵢ
//   ddTable(i, j) = f[xᵢ₋ⱼ, ..., xᵢ] for j >= 1
//
// the diagonal elements ddTable(k, k) are NOT the polynomial coefficients —
// the coefficients are ddTable(j, j) which equals f[x₀, x₁, ..., xⱼ]
void NewtonDividedDifference::buildTable() const {
  if (computed) return;

  if (numPoints < 2)
    throw MatrixException(
        "Newton divided difference needs at least 2 data points");

  // check for duplicate x values — would cause division by zero
  for (int i = 0; i < numPoints; i++) {
    for (int j = i + 1; j < numPoints; j++) {
      if (fabs(getX(i) - getX(j)) < 1e-10)
        throw MatrixException(
            "duplicate x values found... Newton divided difference can't "
            "handle that (division by zero)");
    }
  }

  int n = numPoints;

  // create n×n table initialized to 0
  ddTable = Matrix(n, n);

  // column 0: zeroth-order divided differences = y values
  for (int i = 0; i < n; i++) {
    ddTable.setData(i, 0, getY(i));
  }

  // fill columns 1 to n-1: higher-order divided differences
  for (int j = 1; j < n; j++) {
    for (int i = j; i < n; i++) {
      double num = ddTable.getData(i, j - 1) - ddTable.getData(i - 1, j - 1);
      double den = getX(i) - getX(i - j);
      ddTable.setData(i, j, num / den);
    }
  }

  computed = true;
}

// evaluate the Newton polynomial at a single x using Horner's nested form
//
// P(x) = f[x₀] + (x-x₀)(f[x₀,x₁] + (x-x₁)(f[x₀,x₁,x₂] + ...))
//
// working from the inside out (highest order to lowest):
//   result = coeff[n-1]
//   for i = n-2 down to 0:
//       result = result * (x - x[i]) + coeff[i]
double NewtonDividedDifference::evaluate(double x) const {
  if (!computed) buildTable();

  int n = numPoints;

  // coefficients are the diagonal elements: ddTable(k, k) = f[x₀, ..., xₖ]
  // Horner's form: start from the last coefficient and work backwards
  double result = ddTable.getData(n - 1, n - 1);

  for (int i = n - 2; i >= 0; i--) {
    result = result * (x - getX(i)) + ddTable.getData(i, i);
  }

  return result;
}

// print the divided difference table and polynomial coefficients
void NewtonDividedDifference::printFitInfo() {
  if (!computed) buildTable();

  int n = numPoints;

  cout << fixed << setprecision(6);
  cout << "\n--- Newton Divided Difference Table ---" << endl;

  // header row
  cout << left << setw(12) << "x";
  cout << setw(14) << "f[.]";
  for (int j = 1; j < n; j++) {
    string header = "f[";
    for (int k = 0; k <= j; k++) {
      if (k > 0) header += ",";
      header += ".";
    }
    header += "]";
    cout << setw(14) << header;
  }
  cout << endl;
  cout << string(12 + 14 * n, '-') << endl;

  // table rows
  for (int i = 0; i < n; i++) {
    cout << left << setw(12) << getX(i);
    for (int j = 0; j <= i; j++) {
      cout << setw(14) << ddTable.getData(i, j);
    }
    cout << endl;
  }

  // polynomial coefficients (diagonal elements)
  cout << "\n--- Polynomial Coefficients ---" << endl;
  for (int k = 0; k < n; k++) {
    cout << "  f[x₀";
    for (int j = 1; j <= k; j++) {
      cout << ", x" << j;
    }
    cout << "] = " << ddTable.getData(k, k) << endl;
  }

  // polynomial expression
  cout << "\n  P(x) = " << ddTable.getData(0, 0);
  for (int k = 1; k < n; k++) {
    double coeff = ddTable.getData(k, k);
    if (coeff >= 0)
      cout << " + " << coeff;
    else
      cout << " - " << fabs(coeff);
    for (int j = 0; j < k; j++) {
      double xj = getX(j);
      if (xj >= 0)
        cout << "(x - " << xj << ")";
      else
        cout << "(x + " << fabs(xj) << ")";
    }
  }
  cout << endl;
}

// get a specific divided difference coefficient f[x₀, ..., xₖ]
double NewtonDividedDifference::getCoefficient(int k) const {
  if (!computed) buildTable();

  if (k < 0 || k >= numPoints)
    throw MatrixException("coefficient index out of bounds in getCoefficient()");

  return ddTable.getData(k, k);
}

// get the full divided difference table as a Matrix
Matrix NewtonDividedDifference::getTable() const {
  if (!computed) buildTable();
  return ddTable;
}

File: Extended_in_python/pynumerics/interpolation/newton_divided_difference.py
"""
Newton's Divided Difference Interpolation.

P(x) = f[x₀] + f[x₀,x₁](x−x₀) + f[x₀,x₁,x₂](x−x₀)(x−x₁) + ⋯

The divided difference table is built recursively:
  f[xᵢ]           = yᵢ
  f[xᵢ, xᵢ₊₁]    = (f[xᵢ₊₁] − f[xᵢ]) / (xᵢ₊₁ − xᵢ)
  f[xᵢ, …, xᵢ₊ₖ] = (f[xᵢ₊₁,…,xᵢ₊ₖ] − f[xᵢ,…,xᵢ₊ₖ₋₁]) / (xᵢ₊ₖ − xᵢ)
"""

from pynumerics.interpolation.base import Interpolation
from pynumerics.exceptions import MatrixException


class NewtonDividedDifference(Interpolation):
    """Newton's Divided Difference polynomial interpolation."""

    def __init__(self, x_data: list[float], y_data: list[float]):
        super().__init__(x_data, y_data)
        self._dd_table: list[list[float]] | None = None
        self._coefficients: list[float] | None = None

    def _build_table(self) -> None:
        """Build the divided difference table (lazy, computed once)."""
        if self._dd_table is not None:
            return

        if self.num_points < 2:
            raise MatrixException(
                "Newton divided difference needs at least 2 data points"
            )

        # check for duplicate x values
        for i in range(self.num_points):
            for j in range(i + 1, self.num_points):
                if abs(self._x[i] - self._x[j]) < 1e-10:
                    raise MatrixException(
                        "duplicate x values found... Newton divided difference "
                        "can't handle that (division by zero)"
                    )

        n = self.num_points

        # build n×n table (lower-triangular)
        # table[i][j] = f[xᵢ₋ⱼ, ..., xᵢ]
        table = [[0.0] * n for _ in range(n)]

        # column 0: zeroth-order divided differences = y values
        for i in range(n):
            table[i][0] = self._y[i]

        # fill columns 1 to n-1
        for j in range(1, n):
            for i in range(j, n):
                num = table[i][j - 1] - table[i - 1][j - 1]
                den = self._x[i] - self._x[i - j]
                table[i][j] = num / den

        self._dd_table = table

        # extract diagonal coefficients: table[k][k] = f[x₀, ..., xₖ]
        self._coefficients = [table[k][k] for k in range(n)]

    def evaluate(self, x: float) -> float:
        """Evaluate the Newton polynomial at x using Horner's nested form.

        P(x) = f[x₀] + (x−x₀)(f[x₀,x₁] + (x−x₁)(f[x₀,x₁,x₂] + ...))

        Working from inside out:
            result = coeff[n-1]
            for i = n-2 down to 0:
                result = result * (x - x[i]) + coeff[i]
        """
        self._build_table()

        n = self.num_points
        coeffs = self._coefficients

        # Horner's form
        result = coeffs[n - 1]
        for i in range(n - 2, -1, -1):
            result = result * (x - self._x[i]) + coeffs[i]

        return result

    def get_table(self) -> list[list[float]]:
        """Return the full divided difference table."""
        self._build_table()
        return self._dd_table

    def get_coefficients(self) -> list[float]:
        """Return the diagonal coefficients [f[x₀], f[x₀,x₁], ...]."""
        self._build_table()
        return list(self._coefficients)

    def print_table(self) -> None:
        """Print the divided difference table in a formatted way."""
        self._build_table()

        n = self.num_points
        table = self._dd_table

        print("\n--- Newton Divided Difference Table ---")

        # header row
        header = f"{'x':<12}{'f[.]':<14}"
        for j in range(1, n):
            dots = ",".join(["."] * (j + 1))
            header += f"{'f[' + dots + ']':<14}"
        print(header)
        print("-" * (12 + 14 * n))

        # table rows
        for i in range(n):
            row = f"{self._x[i]:<12.6f}"
            for j in range(i + 1):
                row += f"{table[i][j]:<14.6f}"
            print(row)

        # polynomial coefficients
        print("\n--- Polynomial Coefficients ---")
        for k in range(n):
            label = "f[x₀" + "".join(f", x{j}" for j in range(1, k + 1)) + "]"
            print(f"  {label} = {self._coefficients[k]:.6f}")

        # polynomial expression
        expr = f"\n  P(x) = {self._coefficients[0]:.6f}"
        for k in range(1, n):
            coeff = self._coefficients[k]
            if coeff >= 0:
                expr += f" + {coeff:.6f}"
            else:
                expr += f" - {abs(coeff):.6f}"
            for j in range(k):
                xj = self._x[j]
                if xj >= 0:
                    expr += f"(x - {xj:.6f})"
                else:
                    expr += f"(x + {abs(xj):.6f})"
        print(expr)

    def plot(self, num_samples: int = 200, title: str | None = None,
             show_points: bool = True, save_path: str | None = None,
             func=None, func_label: str | None = None) -> None:
        """Plot with Newton-specific title."""
        if title is None:
            title = "Newton Divided Difference Interpolation"
        super().plot(
            num_samples=num_samples, title=title,
            show_points=show_points, save_path=save_path,
            func=func, func_label=func_label,
        )
